File: Strategy/Strategy.php
<?php
/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 17:31
 */

interface Strategy {
    public function doOperation($num1,$num2);
}


class OperationAdd implements Strategy{

    public function doOperation($num1, $num2)
    {
        // TODO: Implement doOperation() method.
        return $num1+$num2;
    }
}



class OperationSubstract implements Strategy{

    public function doOperation($num1, $num2)
    {
        // TODO: Implement doOperation() method.
        return $num1-$num2;
    }
}


class OperationMultiply implements Strategy{

    public function doOperation($num1, $num2)
    {
        // TODO: Implement doOperation() method.
        return $num1*$num2;
    }
}

class Context{
    private $strategy;
    public function __construct($strategy)
    {
        $this->strategy=$strategy;
    }

    public function executeStrategy($num1,$num2){
        return $this->strategy->doOperation($num1,$num2);
    }
}

//demo
$context=new Context(new OperationAdd());
var_dump("10+5=".$context->executeStrategy(10,5));

$context=new Context(new OperationSubstract());
var_dump("10-5=".$context->executeStrategy(10,5));

$context=new Context(new OperationMultiply());
var_dump("10*5=".$context->executeStrategy(10,5));

File: Strategy/Strategy.go
package main

import "fmt"

type Strategy interface {
	doOperation(num1,num2 int) int
}

type OperationAdd struct {

}

func(o *OperationAdd)doOperation(num1,num2 int) int{
	return num1+num2
}

type OperationSubstract struct {

}

func(o *OperationSubstract)doOperation(num1,num2 int) int{
	return num1-num2
}


type OperationMultiply struct {

}

func(o *OperationMultiply)doOperation(num1,num2 int) int{
	return num1*num2
}


type Context struct {
	strategy Strategy
}

func(c *Context)executeStrategy(num1,num2 int)int{
	return c.strategy.doOperation(num1,num2)
}


func main(){
	context:=&Context{new(OperationAdd)}
	fmt.Println("10+5=",context.executeStrategy(10,5))

	context=&Context{new(OperationSubstract)}
	fmt.Println("10-5=",context.executeStrategy(10,5))

	context=&Context{new(OperationMultiply)}
	fmt.Println("10*5=",context.executeStrategy(10,5))

}